weave/trace_server_bindings/async_batch_processor.py

```diff
@@ -0,0 +1,111 @@
+import atexit
+import logging
+import time
+from queue import Empty, Full, Queue
+from threading import Event, Lock, Thread
+from typing import Callable, Generic, TypeVar
+
+from weave.trace.context.tests_context import get_raise_on_captured_errors
+
+T = TypeVar("T")
+logger = logging.getLogger(__name__)
+
+
+class AsyncBatchProcessor(Generic[T]):
+    """A class that asynchronously processes batches of items using a provided processor function."""
+
+    def __init__(
+        self,
+        processor_fn: Callable[[list[T]], None],
+        max_batch_size: int = 100,
+        min_batch_interval: float = 1.0,
+        max_queue_size: int = 10_000,
+    ) -> None:
+        """
+        Initializes an instance of AsyncBatchProcessor.
+
+        Args:
+            processor_fn (Callable[[list[T]], None]): The function to process the batches of items.
+            max_batch_size (int, optional): The maximum size of each batch. Defaults to 100.
+            min_batch_interval (float, optional): The minimum interval between processing batches. Defaults to 1.0.
+            max_queue_size (int, optional): The maximum number of items to hold in the queue. Defaults to 10_000.  0 means no limit.
+        """
+        self.processor_fn = processor_fn
+        self.max_batch_size = max_batch_size
+        self.min_batch_interval = min_batch_interval
+        self.queue: Queue[T] = Queue(maxsize=max_queue_size)
+        self.lock = Lock()
+        self.stop_accepting_work_event = Event()
+        self.processing_thread = Thread(target=self._process_batches)
+        self.processing_thread.daemon = True
+        self.processing_thread.start()
+
+        # TODO: Probably should include a health check thread here.  It will revive the
```

Comment:
Per discussion with Jeff, it sounds like this monitoring thread won't be necessary, but I'm leaving this comment here in case we want to revisit in future :)